src/platform/macos/window.rs

```diff
@@ -0,0 +1,634 @@
+                let state: *mut c_void = *this.get_ivar("winitState");
+                let state = &mut *(state as *mut DelegateState);
+                emit_event(state, WindowEvent::Closed);
+            }
+            YES
+        }
+
+        extern fn window_did_resize(this: &Object, _: Sel, _: id) {
+            unsafe {
+                let state: *mut c_void = *this.get_ivar("winitState");
+                let state = &mut *(state as *mut DelegateState);
+                let rect = NSView::frame(*state.view);
+                let scale_factor = NSWindow::backingScaleFactor(*state.window) as f32;
+                let width = (scale_factor * rect.size.width as f32) as u32;
+                let height = (scale_factor * rect.size.height as f32) as u32;
+                emit_event(state, WindowEvent::Resized(width, height));
+            }
+        }
+
+        extern fn window_did_become_key(this: &Object, _: Sel, _: id) {
+            unsafe {
+                // TODO: center the cursor if the window had mouse grab when it
+                // lost focus
+                let state: *mut c_void = *this.get_ivar("winitState");
+                let state = &mut *(state as *mut DelegateState);
+                emit_event(state, WindowEvent::Focused(true));
+            }
+        }
+
+        extern fn window_did_resign_key(this: &Object, _: Sel, _: id) {
+            unsafe {
+                let state: *mut c_void = *this.get_ivar("winitState");
+                let state = &mut *(state as *mut DelegateState);
+                emit_event(state, WindowEvent::Focused(false));
+            }
+        }
+
+        static mut DELEGATE_CLASS: *const Class = 0 as *const Class;
+        static INIT: std::sync::Once = std::sync::ONCE_INIT;
+
+        INIT.call_once(|| unsafe {
+            // Create new NSWindowDelegate
+            let superclass = Class::get("NSObject").unwrap();
+            let mut decl = ClassDecl::new("WinitWindowDelegate", superclass).unwrap();
+
+            // Add callback methods
+            decl.add_method(sel!(windowShouldClose:),
+                window_should_close as extern fn(&Object, Sel, id) -> BOOL);
+            decl.add_method(sel!(windowDidResize:),
+                window_did_resize as extern fn(&Object, Sel, id));
+
+            decl.add_method(sel!(windowDidBecomeKey:),
+                window_did_become_key as extern fn(&Object, Sel, id));
+            decl.add_method(sel!(windowDidResignKey:),
+                window_did_resign_key as extern fn(&Object, Sel, id));
+
+            // Store internal state as user data
+            decl.add_ivar::<*mut c_void>("winitState");
+
+            DELEGATE_CLASS = decl.register();
+        });
+
+        unsafe {
+            DELEGATE_CLASS
+        }
+    }
+
+    fn new(state: DelegateState) -> WindowDelegate {
+        // Box the state so we can give a pointer to it
+        let mut state = Box::new(state);
+        let state_ptr: *mut DelegateState = &mut *state;
```

Comment:
Just noticed this in some of the old code. This should probably use `Box::into_raw`, otherwise I'm pretty sure `state` gets `free`d at the end of this scope when `drop` is called on the `Box`. Seems to me that this has just been "working" by pure chance that the memory at the address hasn't been over-written?